File: apps/smahties/src/vector.rs
pub fn vector_to_blob(vector: &[f32]) -> Vec<u8> {
    let mut blob = Vec::with_capacity(size_of_val(vector));
    for value in vector {
        blob.extend_from_slice(&value.to_le_bytes());
    }
    blob
}

pub fn vector_from_blob(blob: &[u8]) -> Result<Vec<f32>, String> {
    if !blob.len().is_multiple_of(size_of::<f32>()) {
        return Err("embedding vector blob length is not a multiple of f32 size".into());
    }

    Ok(blob
        .chunks_exact(size_of::<f32>())
        .map(|chunk| f32::from_le_bytes([chunk[0], chunk[1], chunk[2], chunk[3]]))
        .collect())
}

pub fn cosine_similarity(left: &[f32], right: &[f32]) -> Option<f32> {
    if left.len() != right.len() || left.is_empty() {
        return None;
    }

    let mut dot = 0.0f32;
    let mut left_norm = 0.0f32;
    let mut right_norm = 0.0f32;

    for (left, right) in left.iter().zip(right) {
        dot += left * right;
        left_norm += left * left;
        right_norm += right * right;
    }

    if left_norm == 0.0 || right_norm == 0.0 {
        return None;
    }

    Some(dot / (left_norm.sqrt() * right_norm.sqrt()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn vector_blob_round_trips() {
        let vector = vec![1.0, -2.5, 3.25];

        assert_eq!(vector_from_blob(&vector_to_blob(&vector)).unwrap(), vector);
    }

    #[test]
    fn cosine_similarity_scores_identical_vectors_as_one() {
        let score = cosine_similarity(&[1.0, 2.0, 3.0], &[1.0, 2.0, 3.0]).unwrap();

        assert!((score - 1.0).abs() < f32::EPSILON);
    }

    #[test]
    fn cosine_similarity_rejects_dimension_mismatch() {
        assert_eq!(cosine_similarity(&[1.0], &[1.0, 2.0]), None);
    }
}

File: src/smahties/vector.py
from __future__ import annotations

import math
import struct


def vector_to_blob(vector: list[float]) -> bytes:
    return b"".join(struct.pack("<f", value) for value in vector)


def vector_from_blob(blob: bytes) -> list[float]:
    if len(blob) % 4 != 0:
        raise ValueError("embedding vector blob length is not a multiple of f32 size")
    return [item[0] for item in struct.iter_unpack("<f", blob)]


def vector_norm(vector: list[float]) -> float:
    return math.sqrt(sum(value * value for value in vector))


def cosine_similarity(left: list[float], right: list[float]) -> float | None:
    return cosine_similarity_with_norms(left, right, vector_norm(left), vector_norm(right))


def cosine_similarity_with_norms(
    left: list[float],
    right: list[float],
    left_norm: float,
    right_norm: float,
) -> float | None:
    if len(left) != len(right) or not left or left_norm == 0.0 or right_norm == 0.0:
        return None
    return sum(a * b for a, b in zip(left, right, strict=True)) / (left_norm * right_norm)
